En esta jupyter notebook se desarrolla la extracción, transformación y carga del conjunto de datos para 2019.

In [12]:
import pandas as pd
import json

%load_ext autoreload
%autoreload 2

import utils

import warnings
warnings.filterwarnings("ignore")

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


Se extraen los datos desde el archivo .csv, se convierte en Dataframe y se observa su contenido.

In [13]:
# Ruta al dataset
ruta_siniestro2023 = '../Datasets/feu-siniestros - (FC)feu-siniestros-2023 (1).csv'

# Leer el archivo CSV directamente en un DataFrame
df_sin2023 = pd.read_csv(ruta_siniestro2023)

# Mostrar el DataFrame
df_sin2023

id_feu    unidad_regional siniestro_fecha siniestro_hora   provincia  \
0    182665  OBSERVATORIO VIAL      2023-12-30       12:30:00  CORRIENTES   
1    197455  OBSERVATORIO VIAL      2023-12-27        0:15:00  CORRIENTES   
2    207065  OBSERVATORIO VIAL      2023-12-27       10:10:00  CORRIENTES   
3    202001  OBSERVATORIO VIAL      2023-12-24       13:50:00  CORRIENTES   
4    207724  OBSERVATORIO VIAL      2023-12-24       23:45:00  CORRIENTES   
..      ...                ...             ...            ...         ...   
379  185920  OBSERVATORIO VIAL      2023-01-07        6:25:00  CORRIENTES   
380  184878  OBSERVATORIO VIAL      2023-01-04        9:10:00  CORRIENTES   
381  189219  OBSERVATORIO VIAL      2023-01-03       10:40:00  CORRIENTES   
382  185890  OBSERVATORIO VIAL      2023-01-03       19:10:00  CORRIENTES   
383  199928  OBSERVATORIO VIAL      2023-01-03        8:40:00  CORRIENTES   

     id_provincia departamento  id_departamento   localidad zona_ocurrencia  \
0              18      CAPITAL               21  CORRIENTES          Urbana   
1              18      CAPITAL               21  CORRIENTES          Urbana   
2              18      CAPITAL               21  CORRIENTES          Urbana   
3              18      CAPITAL               21  CORRIENTES          Urbana   
4              18      CAPITAL               21  CORRIENTES          Urbana   
..            ...          ...              ...         ...             ...   
379            18      CAPITAL               21  CORRIENTES          Urbana   
380            18      CAPITAL               21  CORRIENTES          Urbana   
381            18      CAPITAL               21  CORRIENTES          Urbana   
382            18      CAPITAL               21  CORRIENTES          Urbana   
383            18      CAPITAL               21  CORRIENTES          Urbana   

     ... senalizacion_carteleria senalizacion_horizontal  \
0    ...                      Si                      Si   
1    ...                     S/D                     S/D   
2    ...                      Si                      Si   
3    ...                     S/D                     S/D   
4    ...                     S/D                     S/D   
..   ...                     ...                     ...   
379  ...                      Si                      Si   
380  ...                     S/D                     S/D   
381  ...                     S/D                     S/D   
382  ...                      Si                      Si   
383  ...                     S/D                     S/D   

    senalizacion_transitoria luminosidad luz_artificial transito_restringido  \
0                         No         Día            NaN                  NaN   
1                        S/D       Noche             SI                  NaN   
2                         No         Día            NaN                  NaN   
3                        S/D         Día            NaN                  NaN   
4                        S/D       Noche             NO                  NaN   
..                       ...         ...            ...                  ...   
379                       Si       Noche            NaN                  NaN   
380                      S/D         Día            NaN                  NaN   
381                      S/D         Día            NaN                  NaN   
382                       Si       Noche             SI                  NaN   
383                       Si         Día            NaN                  NaN   

    estado_ambiental  visibilidad  visibilidad_otro      semaforo  
0          Despejado          NaN               NaN      Funciona  
1          Despejado          NaN               NaN      Funciona  
2          Despejado          NaN               NaN  Sin Semáforo  
3          Despejado          NaN               NaN  Sin Semáforo  
4          Despejado          NaN               NaN  Sin Semáforo  
..               ...          ...        

Se revisan los tipos de datos por columna y la cantidad de nulos.

In [14]:
utils.verificar_tipo_datos(df_sin2023)

nombre_campo                        tipo_datos  \
0                              id_feu                   [<class 'int'>]   
1                     unidad_regional                   [<class 'str'>]   
2                     siniestro_fecha                   [<class 'str'>]   
3                      siniestro_hora                   [<class 'str'>]   
4                           provincia                   [<class 'str'>]   
5                        id_provincia                   [<class 'int'>]   
6                        departamento                   [<class 'str'>]   
7                     id_departamento                   [<class 'int'>]   
8                           localidad                   [<class 'str'>]   
9                     zona_ocurrencia                   [<class 'str'>]   
10                        via_publica                   [<class 'str'>]   
11                         nombre_via  [<class 'str'>, <class 'float'>]   
12                          altura_km  [<class 'str'>, <class 'float'>]   
13                      entre_calle_1  [<class 'str'>, <class 'float'>]   
14                      entre_calle_2  [<class 'float'>, <class 'str'>]   
15                            latitud  [<class 'float'>, <class 'str'>]   
16                           longitud  [<class 'float'>, <class 'str'>]   
17                             ilesos                 [<class 'float'>]   
18                            heridos                 [<class 'float'>]   
19                         fallecidos                 [<class 'float'>]   
20                          vehiculos                 [<class 'float'>]   
21                           peatones                 [<class 'float'>]   
22            tipo_siniestro_multiple  [<class 'str'>, <class 'float'>]   
23               tipo_siniestro_unico  [<class 'str'>, <class 'float'>]   
24                    despiste_previo  [<class 'float'>, <class 'str'>]   
25                tipo_siniestro_otro  [<class 'float'>, <class 'str'>]   
26                      tipo_colision  [<class 'str'>, <class 'float'>]   
27           cantidad_de_involucrados                   [<class 'int'>]   
28                        trazado_via                 [<class 'float'>]   
29            configuracion_de_la_via  [<class 'str'>, <class 'float'>]   
30  configuracion_de_la_via_otro_tipo  [<class 'float'>, <class 'str'>]   
31                  pendiente_declive                 [<class 'float'>]   
32                     cono_de_vision  [<class 'str'>, <class 'float'>]   
33             material_de_la_calzada  [<class 'str'>, <class 'float'>]   
34        material_de_la_calzada_otro  [<class 'float'>, <class 'str'>]   
35               estado_de_la_calzada  [<class 'str'>, <class 'float'>]   
36            estado_fisico_ambiental  [<class 'str'>, <class 'float'>]   
37        estado_de_conservacion_otro                 [<class 'float'>]   
38          division_fisica_de_la_via  [<class 'str'>, <class 'float'>]   
39     division_fisica_de_la_via_otro  [<class 'float'>, <class 'str'>]   
40              barreras_de_seguridad  [<class 'str'>, <class 'float'>]   
41                          banquinas  [<class 'float'>, <class 'str'>]   
42            senalizacion_carteleria  [<class 'str'>, <class 'float'>]   
43            senalizacion_horizontal  [<class 'str'>, <class 'float'>]   
44           senalizacion_transitoria  [<class 'str'>, <class 'float'>]   
45                        luminosidad  [<class 'str'>, <class 'float'>]   
46                     luz_artificial  [<class 'float'>, <class 'str'>]   
47               transito_restringido                 [<class 'float'>]   
48                   estado_ambiental  [<class 'str'>, <class 'float'>]   
49                        visibilidad                 [<class 'float'>]   
50                   visibilidad_otro                 [<class 'float'>]   
51                           semaforo  [<class 'str'>, <class 'float'>]   

    no_nulos_%  nulos_%  nulos  
0       100.00  

Se observa que hay columnas con más de 90% de nulos, por lo cual decidimos borrarlas.

In [15]:
# Eliminar columnas con más del 90% de valores nulos
umbral = 0.9  # Umbral del 90%
df_2023filtrado = df_sin2023.dropna(thresh=len(df_sin2023) * (1 - umbral), axis=1)

# Mostrar las columnas que se mantuvieron
df_2023filtrado.columns

Index(['id_feu', 'unidad_regional', 'siniestro_fecha', 'siniestro_hora',
       'provincia', 'id_provincia', 'departamento', 'id_departamento',
       'localidad', 'zona_ocurrencia', 'via_publica', 'nombre_via',
       'altura_km', 'entre_calle_1', 'latitud', 'longitud', 'ilesos',
       'heridos', 'fallecidos', 'vehiculos', 'peatones',
       'tipo_siniestro_multiple', 'tipo_siniestro_unico', 'tipo_colision',
       'cantidad_de_involucrados', 'configuracion_de_la_via', 'cono_de_vision',
       'material_de_la_calzada', 'estado_de_la_calzada',
       'estado_fisico_ambiental', 'division_fisica_de_la_via',
       'barreras_de_seguridad', 'senalizacion_carteleria',
       'senalizacion_horizontal', 'senalizacion_transitoria', 'luminosidad',
       'luz_artificial', 'estado_ambiental', 'semaforo'],
      dtype='object')

Verificamos ahora duplicados

In [16]:
utils.verifica_duplicados_por_columna(df_2023filtrado, 'id_feu')

'No hay duplicados'

No hay duplicados, podemos continuar corrigiendo los tipos de datos en algunas columnas. Por ejemplo, fechas y horas.

In [17]:
# Convertir la columna siniestro_fecha a tipo datetime
df_2023filtrado['siniestro_fecha'] = pd.to_datetime(df_2023filtrado['siniestro_fecha'], format='%Y-%m-%d', errors='coerce')

# Convertir la columna siniestro_hora a tipo time (si está en formato 24 horas)
df_2023filtrado['siniestro_hora'] = pd.to_datetime(df_2023filtrado['siniestro_hora'], format='%H:%M:%S', errors='coerce').dt.time

# Verificar las conversiones
df_2023filtrado[['siniestro_fecha', 'siniestro_hora']].head()

siniestro_fecha siniestro_hora
0      2023-12-30       12:30:00
1      2023-12-27       00:15:00
2      2023-12-27       10:10:00
3      2023-12-24       13:50:00
4      2023-12-24       23:45:00

Ahora, podemos pasar en limpio qué datos contiene este Dataframe

In [18]:
utils.verificar_tipo_datos(df_2023filtrado)

nombre_campo  \
0                      id_feu   
1             unidad_regional   
2             siniestro_fecha   
3              siniestro_hora   
4                   provincia   
5                id_provincia   
6                departamento   
7             id_departamento   
8                   localidad   
9             zona_ocurrencia   
10                via_publica   
11                 nombre_via   
12                  altura_km   
13              entre_calle_1   
14                    latitud   
15                   longitud   
16                     ilesos   
17                    heridos   
18                 fallecidos   
19                  vehiculos   
20                   peatones   
21    tipo_siniestro_multiple   
22       tipo_siniestro_unico   
23              tipo_colision   
24   cantidad_de_involucrados   
25    configuracion_de_la_via   
26             cono_de_vision   
27     material_de_la_calzada   
28       estado_de_la_calzada   
29    estado_fisico_ambiental   
30  division_fisica_de_la_via   
31      barreras_de_seguridad   
32    senalizacion_carteleria   
33    senalizacion_horizontal   
34   senalizacion_transitoria   
35                luminosidad   
36             luz_artificial   
37           estado_ambiental   
38                   semaforo   

                                           tipo_datos  no_nulos_%  nulos_%  \
0                                     [<class 'int'>]      100.00     0.00   
1                                     [<class 'str'>]      100.00     0.00   
2   [<class 'pandas._libs.tslibs.timestamps.Timest...      100.00     0.00   
3                           [<class 'datetime.time'>]      100.00     0.00   
4                                     [<class 'str'>]      100.00     0.00   
5                                     [<class 'int'>]      100.00     0.00   
6                                     [<class 'str'>]      100.00     0.00   
7                                     [<class 'int'>]      100.00     0.00   
8                                     [<class 'str'>]      100.00     0.00   
9                                     [<class 'str'>]      100.00     0.00   
10                                    [<class 'str'>]      100.00     0.00   
11                   [<class 'str'>, <class 'float'>]       99.74     0.26   
12                   [<class 'str'>, <class 'float'>]       98.70     1.30   
13                   [<class 'str'>, <class 'float'>]       75.78    24.22   
14                   [<class 'float'>, <class 'str'>]       10.94    89.06   
15                   [<class 'float'>, <class 'str'>]       10.94    89.06   
16                                  [<class 'float'>]       99.74     0.26   
17                                  [<class 'float'>]       99.74     0.26   
18                                  [<class 'float'>]       99.74     0.26   
19                                  [<class 'float'>]       99.74     0.26   
20                                  [<class 'float'>]       99.74     0.26   
21                   [<class 'str'>, <class 'float'>]       98.70     1.30   
22                   [<class 'str'>, <class 'float'>]       98.70     1.30   
23                   [<class 'str'>, <class 'float'>]       46.09    53.91   
24                                    [<class 'int'>]      100.00     0.00   
25                   [<class 'str'>, <class 'float'>]       98.70     1.30   
26                   [<class 'str'>, <class 'float'>]       60.68    39.32   
27                   [<class 'str'>, <class 'float'>]       97.40     2.60   
28                   [<class 'str'>, <class 'float'>]       98.70     1.30   
29                   [<class 'str'>, <class 'float'>]       98.70     1.30   
30                   [<class 'str'>, <class 'float'>]       94.27     5.73   
31                   [<class 'str'>, <class 'float'>]       93.75     6.25   
32                   [<class 'str'>, <class 'float'>]       97.92     2.08   
33                   [<class 'str'>, <class 

In [19]:
print(df_2023filtrado.shape)

(384, 39)


Este conjunto contiene 39 columnas y quedó con 384 filas luego de borrar las columnas con mayor porcentaje de nulos. Las columnas que contiene son:

* id_feu: es el índice del siniestro.
* unidad_regional: es el nombre de la unidad regional encargada de cargar el dato. En este caso, el 100% corresponde a Observatorio Vial.
* siniestro_fecha: es la fecha del siniestro.
* siniestro_hora: es la hora del siniestro.
* id_provincia: es la provincia en la que ocurrió el siniestro. En nuestro caso, el 100% corresponde a Corrientes.
* departamento: es el departamento en el que ocurrió el siniestro. En nuestro caso, el 100% corresponde a Capital.
* id_departamento: es la ID del departamento en el que ocurrió el siniestro.
* localidad: es la localidad en la que ocurrió el siniestro. En nuestro caso, el 100% corresponde a Corrientes.
* zona_ocurrencia: es la especificación sobre el tipo de zona donde ocurrió el siniestro. Urbana o Rural.
* via_publica: es la especificación sobre el tipo de calle en la que ocurrió el siniestro. Avenida, Calle o Ruta.
* nombre_via: es el nombre de la calle, avenida o ruta donde ocurrió el siniestro.
* altura_km: es la altura en kilómetros dónde ocurrió el siniestro.
* entre_calle_1: es el cruce en donde ocurrió el siniestro.
* latitud: es la latitud de dónde ocurrió el siniestro.
* longitud: es la longitud de dónde ocurrió el siniestro.
* ilesos: es la cantidad de personas que no sufrieron daños en el siniestro.
* heridos: es la cantidad de personas que sufrieron daños en el siniestro.
* fallecidos: es la cantidad de personas que fallecieron en el siniestro.
* vehiculos: es la cantidad de vehículos involucrados en el siniestro.
* peatones: es la cantidad de peatones involucrados en el siniestro.
* tipo_siniestro_multiple: es la especificación del hecho.
* despiste_previo: es la respuesta ante el hecho.
* tipo_colision: es el tipo de colisión del siniestro.
* cantidad_involucrados: es la cantidad de personas que participaron del siniestro.
* trazado_via: es la especficidad sobre el tipo de calle en que ocurrió el hecho. Recta o Curva.
* configuracion_de_la_via: es el tipo de configuración de la vía. Intersección, Cruce, Rotonda.
* pendiente_declive: es si existe pendiente.
* material_de_la_calzada: es el material del que está hecha la calzada donde ocurrió el siniestro.
* estado_de_la_calzada: es el estado en que se encontraba la calzada en el momento del siniestro.
* estado_fisico_ambiental: el estado de la calzada al momento del siniestro. Seca, Mojada.
* division_fisica_de_la_via: es el qué divide la vía donde ocurrió el siniestro. Por ejemplo, barrera de hormigón.
* senalización_carteleria: es si existía señalización en la zona al momento del hecho.
* senalización_horizontal: es si existía señalización horizontal en la zona al momento del hecho.
* senalización_transitoria: es si existía señalización transitoria en la zona al momento del hecho.
* luminosidad: es si el hecho ocurrió de día o de noche. Luz Natural.
* luz_artifical: es si en el lugar del hecho existía luz artificial.
* transito_restringido: es si el lugar era de tránsito restringido.
* estado_ambiental: es cómo estaba el tiempo al momento del hecho. Despejado, lluvioso, etc.
* visibilidad: es cómo se encontraba la visibilidad al momento del hecho. 
* semaforo: es si existe semáforo en la zona del hecho, y su funcionamiento.

In [20]:
# Eliminar columnas innecesarias
columnas_a_eliminar = [
    'unidad_regional',
    'id_provincia',
    'id_departamento',
    'altura_km',
    'tipo_siniestro_multiple',
    'configuracion_de_la_via',
    'division_fisica_de_la_via',
    'senalizacion_carteleria',
    'senalizacion_horizontal',
    'senalizacion_transitoria', 
    'provincia',
    'departamento',
    'localidad',
    ]

df_2023filtrado = df_2023filtrado.drop(columns=columnas_a_eliminar)


Ahora tenemos un Dataframe mucho más manejable, con los datos más importantes para nuestra investigación. Vamos a guardarlo para su futuro uso.

In [21]:
sin2023 = '../Datasets_limpios/siniestros2023.csv'
df_2023filtrado.to_csv(sin2023)